### ISSUE TITLE
Allow `respond_to` to narrow types.

### ISSUE THREAD

**Issue by alexander-stripe** [sorbet/sorbet#3469]
### Problem

Today, sorbet does not reflect knowledge gained from a `respond_to?` call when narrowing types. For instance (this doesn't type-check today):

```ruby
# typed: true
class X < T::Struct
  const :a, Integer
end

class Y < T::Struct
  const :a, Integer
  const :b, String
end

def main
  thing = T.let(Y.new(a: 1, b: "a"), T.any(X, Y))

  if thing.respond_to?(:b)
    thing.b # This can't be X, since X doesn't have a method called b, so must be Y
  end
end
```

[view in sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20X%20%3C%20T%3A%3AStruct%0A%20%20const%20%3Aa%2C%20Integer%0Aend%0A%0Aclass%20Y%20%3C%20T%3A%3AStruct%0A%20%20const%20%3Aa%2C%20Integer%0A%20%20const%20%3Ab%2C%20String%0Aend%0A%0Adef%20main%0A%20%20thing%20%3D%20T.let(Y.new(a%3A%201%2C%20b%3A%20%22a%22)%2C%20T.any(X%2C%20Y))%0A%0A%20%20if%20thing.respond_to%3F(%3Ab)%0A%20%20%20%20thing.b%0A%20%20end%0Aend)

### Proposed solution

In sorbets flow-sensitive type narrowing, leverage information gained from this shape check to rule out types.

**Comment by jez** [comment]
To be clear: the proposal in this ticket is to allow `respond_to?` to select between one or another known type (say, in a union). 

If Sorbet cannot see that a method should be present, `respond_to?` should not allow Sorbet to think that the method suddenly exists. Because even if `respond_to?` is true, Sorbet has no way to know what the method's arity is or what types it's arguments require.

**Instead**, if we know that there's a union between 4 classes, and two of them `respond_to?` some method and the others don't, Sorbet could be smarter.

It gets a little tricky, so I'm leaving some test cases:

[→ example 1, with modules (sorbet.run)](https://sorbet.run/#%23%20typed%3A%20true%0A%0A%0Amodule%20IFoo1%0A%20%20include%20Kernel%0A%20%20def%20foo%3B%20end%0Aend%0Amodule%20IBar%0A%20%20include%20Kernel%0A%20%20def%20bar%3B%20end%0Aend%0A%0Amodule%20IFoo2%0A%20%20include%20Kernel%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20A%0A%20%20include%20IBar%0A%20%20include%20IFoo2%0Aend%0A%0Aextend%20T%3A%3ASig%0Asig%20%7Bparams(x%3A%20T.any(IFoo1%2C%20IBar)).void%7D%0Adef%20example(x)%0A%20%20if%20x.respond_to%3F(%3Afoo)%0A%20%20%20%20T.reveal_type(x)%0A%20%20%20%20%23%20%5E%20A.new%20responds%20to%20%3Afoo%2C%20but%20is%20not%20a%20IFoo1%2C%0A%20%20%20%20%23%20so%20we%20CANNOT%20narrow%20the%20type%20to%20IFoo1%20in%20this%20branch%0A%20%20else%0A%20%20%20%20T.reveal_type(x)%0A%20%20%20%20%23%20After%2015%20minutes%2C%20I%20THINK%20we%20can%20narrow%20the%20type%20in%0A%20%20%20%20%23%20this%20branch%20to%20drop%20subtypes%20of%20anything%20known%20to%0A%20%20%20%20%23%20have%20foo%20(so%20drop%20subtypes%20of%20IFoo1%20in%20this%20case)%0A%20%20end%0Aend%0A%0Aexample(A.new))

[→ example 2, with classes (sorbet.run)](https://sorbet.run/#%23%20typed%3A%20true%0A%0Aclass%20ParentNoFoo%0Aend%0A%0Aclass%20ChildAddsFoo%20%3C%20ParentNoFoo%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20AnotherFoo%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aextend%20T%3A%3ASig%0Asig%20%7Bparams(x%3A%20T.any(AnotherFoo%2C%20ParentNoFoo)).void%7D%0Adef%20example(x)%0A%20%20if%20x.respond_to%3F(%3Afoo)%0A%20%20%20%20T.reveal_type(x)%0A%20%20%20%20%23%20%5E%20ChildAddsFoo.new%20responds%20to%20%3Afoo%2C%20but%20is%20not%20a%20AnotherFoo%2C%0A%20%20%20%20%23%20so%20we%20CANNOT%20narrow%20the%20type%20to%20AnotherFoo%20in%20this%20branch%0A%20%20else%0A%20%20%20%20T.reveal_type(x)%0A%20%20%20%20%23%20After%2015%20minutes%2C%20I%20THINK%20we%20can%20narrow%20the%20type%20in%0A%20%20%20%20%23%20this%20branch%20to%20drop%20subtypes%20of%20anything%20known%20to%0A%20%20%20%20%23%20have%20foo%20(so%20drop%20subtypes%20of%20AnotherFoo%20in%20this%20case)%0A%20%20end%0Aend%0A%0Aexample(ChildAddsFoo.new))

Basically, I think that a `respond_to?` check can only yield new knowledge when the condition is false.

**Comment by jez** [comment]
In particular, the original ticket as written is not correct:

```ruby
thing.b # This can't be X, since X doesn't have a method called b, so must be Y
```

`thing.b` actually **CAN** be `X`, since, `X` could have been subclassed, and the subclass could have added `def b; end`.

This hints that the `respond_to?` check might additionally want to check whether a class is `final!`. If the class is `final!`, then we know that the methods we see in the type are the only methods that could be defined on that class.

**Comment by jez** [comment]
^ To clarify on my previous comment:

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Aclass%20HasFoo%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20DoesNotHaveFoo%0Aend%0A%0Aclass%20ChildDoesNotHaveFoo%20%3C%20DoesNotHaveFoo%0A%20%20def%20foo%28x%29%3B%20end%0Aend%0A%0Asig%20%7Bparams%28x%3A%20T.any%28HasFoo%2C%20DoesNotHaveFoo%29%29.void%7D%0Adef%20example%28x%29%0A%20%20if%20x.respond_to%3F%28%3Afoo%29%0A%20%20%20%20x.foo%0A%20%20end%0Aend%0A%0A%23%20raises%20ArgumentError%20when%20called%2C%20because%20this%20class%20DOES%20have%20a%0A%23%20foo%20method%2C%20but%20it%20requires%20an%20extra%20argument.%0Aexample%28ChildDoesNotHaveFoo.new%29)

```ruby
# typed: true
extend T::Sig

class HasFoo
  def foo; end
end

class DoesNotHaveFoo
end

class ChildDoesNotHaveFoo < DoesNotHaveFoo
  def foo(x); end
end

sig {params(x: T.any(HasFoo, DoesNotHaveFoo)).void}
def example(x)
  if x.respond_to?(:foo)
    x.foo
  end
end

# raises ArgumentError when called, because this class DOES have a
# foo method, but it requires an extra argument.
example(ChildDoesNotHaveFoo.new)
```

This is why we couldn't easily do something like the original ticket recommends.

- - - -

An alternative approach which typechecks would be to factor out an explicit interface, and then use `is_a?` on that interface:

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aextend%20T%3A%3ASig%0A%0Amodule%20IFoo%0A%20%20extend%20T%3A%3AHelpers%0A%20%20extend%20T%3A%3ASig%0A%20%20abstract!%0A%20%20sig%20%7Babstract.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20HasFoo%0A%20%20extend%20T%3A%3ASig%0A%20%20include%20IFoo%0A%20%20sig%20%7Boverride.void%7D%0A%20%20def%20foo%3B%20end%0Aend%0A%0Aclass%20DoesNotHaveFoo%0Aend%0A%0Aclass%20ChildDoesNotHaveFoo%20%3C%20DoesNotHaveFoo%0A%20%20%23%20Try%20uncommenting%20to%20see%20the%20static%20error%3A%0A%20%20%23%20includ%20Foo%0A%20%20def%20foo%28x%29%3B%20end%0Aend%0A%0Asig%20%7Bparams%28x%3A%20T.any%28HasFoo%2C%20DoesNotHaveFoo%29%29.void%7D%0Adef%20example%28x%29%0A%20%20if%20x.is_a%3F%28IFoo%29%0A%20%20%20%20x.foo%0A%20%20end%0Aend%0A%0A%23%20Does%20not%20raise%2C%20because%20ChildClassDoesNotHaveFoo%20is%20not%20an%20IFoo%0A%23%20%28also%2C%20if%20we%20tried%20to%20implement%20it%20by%20putting%20%60include%20IFoo%60%0A%23%20inside%20this%20class's%20definition%2C%20we%20would%20get%20an%20incompatible%0A%23%20override%20checking%20error%2C%20because%20there%20are%20too%20many%20params.%29%0Aexample%28ChildDoesNotHaveFoo.new%29)

```ruby
# typed: true
extend T::Sig

module IFoo
  extend T::Helpers
  extend T::Sig
  abstract!
  sig {abstract.void}
  def foo; end
end

class HasFoo
  extend T::Sig
  include IFoo
  sig {override.void}
  def foo; end
end

class DoesNotHaveFoo
end

class ChildDoesNotHaveFoo < DoesNotHaveFoo
  # Try uncommenting to see the static error:
  # includ Foo
  def foo(x); end
end

sig {params(x: T.any(HasFoo, DoesNotHaveFoo)).void}
def example(x)
  if x.is_a?(IFoo)
    x.foo
  end
end

# Does not raise, because ChildClassDoesNotHaveFoo is not an IFoo
# (also, if we tried to implement it by putting `include IFoo`
# inside this class's definition, we would get an incompatible
# override checking error, because there are too many params.)
example(ChildDoesNotHaveFoo.new)
```
